build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/transport/TransportTestExistTask.java

```diff
@@ -0,0 +1,166 @@
+import org.gradle.workers.WorkQueue;
+import org.gradle.workers.WorkerExecutor;
+import org.objectweb.asm.ClassReader;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+/**
+ * Runs LoggerUsageCheck on a set of directories.
+ */
+@CacheableTask
+public abstract class TransportTestExistTask extends PrecommitTask {
+
+    private FileCollection classpath;
+
+    private final ListProperty<FileCollection> classesDirs;
+
+    private ObjectFactory objectFactory;
+    private FileCollection mainSources;
+    private FileCollection testSources;
+    private Configuration compileClasspath;
+    private Configuration testClasspath;
+
+    @Inject
+    public TransportTestExistTask(ObjectFactory objectFactory) {
+        this.classesDirs = objectFactory.listProperty(FileCollection.class);
+        this.objectFactory = objectFactory;
+        setDescription("Runs TransportTestExistTask on output directories of all source sets");
+    }
+
+    @Inject
+    abstract public WorkerExecutor getWorkerExecutor();
+
+    @TaskAction
+    public void runLoggerUsageTask() {
+        WorkQueue workQueue = getWorkerExecutor().noIsolation();
+        workQueue.submit(TransportTestExistWorkAction.class, parameters -> {
+            parameters.getMainSources().setFrom(mainSources);
+            parameters.getTestSources().setFrom(testSources);
+            parameters.getCompileClasspath().setFrom(compileClasspath);
+            parameters.getTestClasspath().setFrom(testClasspath);
+        });
+    }
+
+    public void setMainSources(FileCollection mainSources) {
+        this.mainSources = mainSources;
+    }
+
+    public void setTestSources(FileCollection testSources) {
+        this.testSources = testSources;
+    }
+
+    public void setCompileClasspath(Configuration compileClasspath) {
+        this.compileClasspath = compileClasspath;
+    }
+
+    public void setTestClasspath(Configuration testClasspath) {
+        this.testClasspath = testClasspath;
+    }
+
+    abstract static class TransportTestExistWorkAction implements WorkAction<Parameters> {
+
+        private final Logger logger = Logging.getLogger(TransportTestExistTask.class);
+
+        private final ExecOperations execOperations;
+
+        @Inject
+        public TransportTestExistWorkAction(ExecOperations execOperations) {
+            this.execOperations = execOperations;
+        }
+
+        @Override
+        public void execute() {
+            Set<File> mainClasses = getParameters().getMainSources().getFiles();
+            Set<File> testClasses = getParameters().getTestSources().getFiles();
+            Set<File> compileClassPath = getParameters().getCompileClasspath().getFiles();
+            Set<File> testClassPath = getParameters().getTestClasspath().getFiles();
+
+            // this is always run.. should be cached?
```

Comment:
almost all gradle modules have a dependency on server and test.framework. I have to scan them to find subclasses of Writeable/AbstractWireTestCase. But this is a silly job to repeat that for all modules.

How should this be possibly refactored to cache these results? can I share something between tasks in gradle?

Reply:
another questions. can we assume that the order of running this task is the same as dependencies?
i.e if `:modules:aggregations:transportTestExistCheck ` will be run after `:server:transportTestExistCheck` ?

if yes, then perhaps we could "emit" an additional generated cache file. so that later runs of this task would use it?

Reply:
no you cannot guarantee that order. Only if you would add an explicit task dependency, which is wrong. You could _maybe_ share processed classes in the worker implementation you use but I strongly feel like not doing that from the start and optimise later

Reply:
Why do you need to find subclasses of Writeable/AbstractWireTestCase? Can't you just check if a class is assignable to those?